fractalis/authorization.py

```diff
@@ -0,0 +1,126 @@
+import functools
+from enum import Enum
+from functools import wraps
+import requests
+import jwt
+from flask import request, jsonify
+import logging
+import json
+
+from jwt.algorithms import RSAAlgorithm
+from werkzeug.exceptions import Unauthorized
+
+logger = logging.getLogger(__name__)
+AuthServiceType = Enum('AuthServiceType', 'OIDC TRANSMART', module=__name__)
+
+
+def authorized(f):
+    @wraps(f)
+    def _wrap(*args, **kwargs):
+        try:
+            payload = request.get_json(force=True)
+            auth = payload.get('auth')
+            token = auth.get('token') if len(auth) != 0 else ''
+            if len(token) == 0:
+                error = "No token in the authentication object."
+                logger.error(error)
+                raise Unauthorized(error)
+            decoded_token = jwt.decode(token, verify=False)
+
+            user = validate_user(decoded_token, auth)
+            identity_provider_url = validate_identity_provider_url(decoded_token, auth)
+            auth_service_type = auth.get('authServiceType', '').upper()
+            if auth_service_type == AuthServiceType.OIDC.name:
+                validate_token(token,  auth.get('oidcClientId'), identity_provider_url)
+            else:
+                logger.warning(f'Token not validated. Validation not supported for "{auth_service_type}" service type')
+        except Exception as e:
+            return jsonify({'error': f'Access unauthorized. {str(e)}'}), 403
+        logger.info(f'Connected: {decoded_token.get("email")!r}, user id (sub): {user!r}')
+        return f(*args, **kwargs)
+    return _wrap
+
+
+def validate_identity_provider_url(decoded_token, auth) -> str:
+    """ Checks if the token issuer (iss) matches the identity provider url from authentication object
+    :param decoded_token: decoded user token
+    :param auth: authentication object from the request arguments
+    :return: identity_provider_url or
+             Unauthorized if urls do not match
+    """
+    token_issuer = f'{decoded_token.get("iss")}/protocol/openid-connect'
+    identity_provider_url = auth.get('oidcServerUrl')
+    if token_issuer != identity_provider_url:
+        error = "Token issuer does not match the identity provider url from authentication object parameter."
+        logger.error(error)
+        raise Unauthorized(error)
+    return identity_provider_url
+
+
+def validate_user(decoded_token, auth):
+    """ Checks if the current user (in the token) matches the user from authentication object
+    :param decoded_token: decoded user token
+    :param auth: authentication object from the request arguments
+    :return: user or
+             Unauthorized if users do not match
+    """
+    subject = decoded_token.get('sub')
+    user = auth.get('user')
```

Comment:
currently there is no user specified in the session, so the validation is based on the auth object passed in request parameters. Not sure what is the best place to change the session to have the username specified.

Reply:
I think we want to support a workflow like this:
- User Alice requests data for the first time
  - A new session is created for Alice
  - Data fetched using the ETL, stored in the cache
  - Data is returned
- User Alice requests data again
  - Token is validated and user id matched with the current session user id
  - Data is read from the cache (if present, otherwise ETL)
  - Data is returned
- User Bob requests data
  - Token validated, but user id does not match the session user id
  - A new session is created for Bob
  - etc.

For this we need to store the subject id when creating a new session (`open_session` in [session.py](src/fractalis/session.py)).

Probably we should also configure an authentication type in `app.config` instead of relying on what type is passed by the user.